File: packages/core/src/schemas/execution.ts
import { z } from "zod";
import { CardIdSchema } from "./card.js";

export const ExecutionIdSchema = z.string().brand<"ExecutionId">();

export const ExecutionPhase = z.enum(["plan", "execute"]);

export const ExecutionStatus = z.enum([
  "running",
  "success",
  "failed",
  "cancelled",
]);

export const FileChangeSchema = z.object({
  path: z.string(),
  additions: z.number(),
  deletions: z.number(),
});

export const ExecutionSchema = z.object({
  id: ExecutionIdSchema,
  card_id: CardIdSchema,
  session_id: z.string().nullable(),
  phase: ExecutionPhase.default("plan"),
  started_at: z.string(),
  finished_at: z.string().nullable(),
  status: ExecutionStatus.default("running"),
  output: z.string().default(""),
  cost_usd: z.number().default(0),
  exit_code: z.number().int().nullable(),
  retry_count: z.number().int().default(0),
  pid: z.number().int().nullable(),
  files_changed: z.string().nullable().default(null),
});

File: rust/crates/core/src/types/execution.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
#[serde(rename_all = "lowercase")]
pub enum ExecutionPhase {
    Plan,
    Execute,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS, PartialEq)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
#[serde(rename_all = "lowercase")]
pub enum ExecutionStatus {
    Running,
    Success,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../../packages/dashboard/src/types/generated/")]
pub struct Execution {
    pub id: String,
    pub card_id: String,
    pub session_id: Option<String>,
    pub phase: ExecutionPhase,
    pub started_at: String,
    pub finished_at: Option<String>,
    pub status: ExecutionStatus,
    pub output: String,
    pub cost_usd: f64,
    pub exit_code: Option<i64>,
    pub retry_count: i64,
    pub pid: Option<i64>,
    pub files_changed: Option<String>,
}
